src/Altinn.Notifications.Persistence/Repository/NotificationSummaryRepository.cs

```diff
@@ -36,12 +37,70 @@ public NotificationSummaryRepository(NpgsqlDataSource dataSource, TelemetryClien
     /// <inheritdoc/>
     public async Task<EmailNotificationSummary?> GetEmailSummary(Guid orderId, string creator)
     {
-        bool matchFound = false;
+        var (matchFound, sendersReference, notificationList) = await GetSummary(
+            orderId,
+            creator,
+            _getEmailNotificationsByOrderIdSql,
+            reader => new EmailNotificationWithResult(
+                reader.GetValue<Guid>("alternateid"),
+                new EmailRecipient()
+                {
+                    RecipientId = reader.GetValue<string>("recipientid"),
+                    ToAddress = reader.GetValue<string>("toaddress")
+                },
+                new NotificationResult<EmailNotificationResultType>(
+                    reader.GetValue<EmailNotificationResultType>("result"),
+                    reader.GetValue<DateTime>("resulttime"))));
+
+        if (!matchFound)
+        {
+            return null;
+        }
 
-        List<EmailNotificationWithResult> notificationList = new();
+        return new EmailNotificationSummary(orderId)
+        {
+            SendersReference = sendersReference,
+            Notifications = notificationList.Cast<EmailNotificationWithResult>().ToList()
+        };
+    }
+
+    /// <inheritdoc/>
+    public async Task<SmsNotificationSummary?> GetSmsSummary(Guid orderId, string creator)
+    {
+        var (matchFound, sendersReference, notificationList) = await GetSummary(
+            orderId,
+            creator,
+            _getSmsNotificationsByOrderIdSql,
+            reader => new SmsNotificationWithResult(
+                reader.GetValue<Guid>("alternateid"),
+                new SmsRecipient()
+                {
+                    RecipientId = reader.GetValue<string>("recipientid"),
+                    MobileNumber = reader.GetValue<string>("mobilenumber")
+                },
+                new NotificationResult<SmsNotificationResultType>(
+                    reader.GetValue<SmsNotificationResultType>("result"),
+                    reader.GetValue<DateTime>("resulttime"))));
+
+        if (!matchFound)
+        {
+            return null;
+        }
+
+        return new SmsNotificationSummary(orderId)
+        {
+            SendersReference = sendersReference,
+            Notifications = notificationList.Cast<SmsNotificationWithResult>().ToList()
```

Comment:
Have you tested this without the cast? Looks like the return from GetSummary is of the correct because of clever use of Generics. 

I also tried to add a parameter to `GetSummary` of type `INotificationSummary<T>`. I believe that could work if it wasn't for the fact that the `INotificationSummary.SendersReference` is read only. Could we have used a base class instead of an interface?

You're doing to cool stuff here.

Reply:
I tried to convert it to a base class instead of interface, but end ut with a cast exception when trying to return the final summary class explicitly. Merging feature for now, and will try a bit more on further simplifying the repository logic later today.